### PR TITLE
Optimise program size for firmware examples

### PR DESCRIPTION
**Pull request by cuddlefishie**
Part of #291 

- switches to nightly Rust
  - allows us to use `build_std`
- switches to [`oxalica/rust-overlay`](https://github.com/oxalica/rust-overlay)
  - the mozzilla overlay didn't handle rustup components well
- use `ufmt` instead of `core::fmt`
- separate firmware-libraries (unit-testable on host machine) from examples (need nightly, optimised, cross-compile only)
- use [`elf-limits`](https://github.com/cuddlefishie/elf-limits) in CI to make sure firmware sizes are kept in check

### CODE REVIEW

File: firmware/examples/fdt-read/memory.x [hunk 1]
```diff
@@ -6,12 +6,13 @@ SPDX-License-Identifier: CC0-1.0
...
+  INSTR : ORIGIN = 0x20000000, LENGTH = 64K
 }
 
-REGION_ALIAS("REGION_TEXT", RAM);
-REGION_ALIAS("REGION_RODATA", RAM);
-REGION_ALIAS("REGION_DATA", RAM);
-REGION_ALIAS("REGION_BSS", RAM);
-REGION_ALIAS("REGION_HEAP", RAM);
-REGION_ALIAS("REGION_STACK", RAM);
+REGION_ALIAS("REGION_TEXT", INSTR);
```
Comment (martijnbastiaan) [inline firmware/examples/fdt-read/memory.x:13]: I'm not particularly well-known with the embedded stuff, but why would `REGION_TEXT` live in the `INSTR` memory space, while the other ones live in `DATA`?

**Review by martijnbastiaan — APPROVED** [review]

File: firmware/examples/fdt-read/memory.x (hunk 1 above)
Comment (cuddlefishie) [reply firmware/examples/fdt-read/memory.x:13]: "text" is old terminology for "instructions", if you have an ELF file, the actual instructions will be in a `.text` section.

This separation here is to have some kind of Harvard architecture where the instruction and data buses are separate. All other things will be accessed via the data bus, so it's in the data memory, while instructions are accessed through the instruction bus.

I think eventually we want to merge those two and have a von-Neumann architecture, but all that's needed here would be to change it back to one unified RAM

File: firmware/examples/fdt-read/src/main.rs [hunk 2]
```diff
@@ -46,15 +44,17 @@ fn print_fdt(depth: usize, node: &fdt::node::FdtNode) {
...
             PropValue::TwoIntegers(a, b) => println!(
                 "{} = ({}, {})",
                 prop.name,
-                best_fit_integer_repr(a),
-                best_fit_integer_repr(b)
+                best_fit_integer_repr(a).as_str(),
+                best_fit_integer_repr(b).as_str()
             ),
-            PropValue::Blob(b) => println!("{} = {b:?}", prop.name),
+            PropValue::Blob(b) => println!("{} = {:?}", prop.name, b),
```
[the remark this answers is not in the document]
Comment (cuddlefishie) [reply firmware/examples/fdt-read/src/main.rs:57]: Yes exactly. The `ufmt` uses a procedural macro to generate the code for formatting and it doesn't support everything that `core::fmt` supports (which uses the compiler built-in `format_args!`), but it saves a lot of instructions :sweat_smile:

File: shell.nix [hunk 3]
```diff
@@ -26,22 +26,7 @@ pkgs.mkShell {
...
-      #
-      # XXX: We cannot swich to 1.67 due to 'cargo clippy' throwing an obscure
-      #      error. See: https://github.com/bittide/bittide-hardware/pull/227#issuecomment-1411808302
-      ((pkgs.rustChannelOf { channel = "1.65"; }).rust.override {
-        targets = [
-          "riscv32imc-unknown-none-elf"
-          "x86_64-unknown-linux-gnu"
-        ];
-      })
+      (pkgs.rust-bin.fromRustupToolchainFile ./rust-toolchain.toml)
```
[the remark this answers is not in the document]
Comment (cuddlefishie) [reply shell.nix:29]: Yep! This overlay also works better with the nightly components and all that! :)

**Pull request merged**